app/repositories/slick/implementations/SlickChatsRepository.scala

```diff
@@ -95,71 +105,94 @@ class SlickChatsRepository @Inject() (db: Database)(implicit executionContext: E
-        case ((groupedChatId, maxDate), (chatId, emailId, _, date, _)) =>
-          (chatId, emailId)
-      }
-      .groupBy(_._1)
-      .map {
-        case (chatId, chatEmailQuery) =>
-          (chatId, chatEmailQuery.map { case (chat, emailId) => emailId }.min)
-      }
+    else {
+      val visibleEmailsQuery = getVisibleEmailsQuery(userId, Some(mailbox))
 
-    val chatPreviewQuery = for {
-      (chatId, emailId) <- groupedQuery
-      subject <- ChatsTable.all.filter(_.chatId === chatId).map(_.subject)
-      (emailId, body, date) <- EmailsTable.all.filter(emailRow =>
-        emailRow.chatId === chatId && emailRow.emailId === emailId).map(emailRow =>
-        (emailRow.emailId, emailRow.body.take(PREVIEW_BODY_LENGTH), emailRow.date))
-      addressId <- EmailAddressesTable.all.filter(emailAddressRow =>
-        emailAddressRow.emailId === emailId && emailAddressRow.participantType === from)
-        .map(_.addressId)
-      address <- AddressesTable.all.filter(_.addressId === addressId).map(_.address)
+      val groupedQuery = visibleEmailsQuery
+        .map { case (chatId, emailId, body, date, sent) => (chatId, date) }
+        .groupBy(_._1)
+        .map {
+          case (chatId, chatDateQuery) =>
+            (chatId, chatDateQuery.map { case (chat, date) => date }.max)
+        }
+        .join(visibleEmailsQuery)
+        .on {
+          case ((groupedChatId, maxDate), (chatId, emailId, _, date, _)) =>
+            groupedChatId === chatId && maxDate === date
+        }
+        .map {
+          case ((groupedChatId, maxDate), (chatId, emailId, _, date, _)) =>
+            (chatId, emailId)
+        }
+        .groupBy(_._1)
+        .map {
+          case (chatId, chatEmailQuery) =>
+            (chatId, chatEmailQuery.map { case (chat, emailId) => emailId }.min)
+        }
 
-    } yield (chatId, subject, address, date, body)
+      val chatPreviewQuery = for {
+        (chatId, emailId) <- groupedQuery
+        subject <- ChatsTable.all.filter(_.chatId === chatId).map(_.subject)
+        (emailId, body, date) <- EmailsTable.all.filter(emailRow =>
+          emailRow.chatId === chatId && emailRow.emailId === emailId).map(emailRow =>
+          (emailRow.emailId, emailRow.body.take(PREVIEW_BODY_LENGTH), emailRow.date))
+        addressId <- EmailAddressesTable.all.filter(emailAddressRow =>
+          emailAddressRow.emailId === emailId && emailAddressRow.participantType === from)
+          .map(_.addressId)
+        address <- AddressesTable.all.filter(_.addressId === addressId).map(_.address)
+
+      } yield (chatId, subject, address, date, body)
+
+      chatPreviewQuery
+        .sortBy { case (chatId, subject, address, date, body) => (date.desc, body.asc, address.asc) }
+        .result
+        .map(chatpreviewData => {
+          val fullChatPreview = chatpreviewData.map(ChatPreview.tupled)
+          val totalCount = fullChatPreview.size
+          Some((sliceSequence(fullChatPreview, perPage, page), totalCount,
```

Comment:
You need to use the take and drop methods from slick.
If you have a thousand chats you will put them all in memory.
You always need to limit what you get from the DB.
Define a maximum.